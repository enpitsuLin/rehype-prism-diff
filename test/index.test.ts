import { rehype } from 'rehype'
import dedent from 'dedent'
import rehypePrismDiff, { Options } from '../src/'
import rehypePrism, { RehypePrismOptions } from 'rehype-prism'
import rehypePrismPlus from 'rehype-prism-plus'
import { visit } from 'unist-util-visit'
import type { Options as RehypePrismPlusOptions } from 'rehype-prism-plus/generator'
import type { Element } from 'hast'

const addMeta = (meta?: string) => {
  if (!meta) return
  return (tree: Element) => {
    visit(tree, 'element', (node, index, parent) => {
      if (node.tagName === 'code') {
        node.data = { meta: meta }
      }
    })
  }
}

describe('rehypre-prism plugin', () => {
  const processHtml = (
    html: string,
    setting?: { options?: RehypePrismOptions; optionsDiff?: Options; meta?: string }
  ) => {
    return rehype()
      .data('settings', { fragment: true })
      .use(addMeta, setting?.meta || '')
      .use(rehypePrism, setting?.options)
      .use(rehypePrismDiff, setting?.optionsDiff)
      .processSync(html)
      .toString()
  }

  test('should no diff', () => {
    const result = processHtml(
      dedent`
    <div>
      <pre>
      <code class="language-css">p { color: red }</code>
      </pre>
    </div>
    `
    )
    expect(result).toMatchSnapshot()
  })

  test('should add diff', () => {
    const result = processHtml(
      dedent`
    <div>
      <pre>
      <code class="language-css">p { color: red }</code>
      </pre>
    </div>
    `,
      {
        meta: 'diff'
      }
    )
    expect(result).toMatchSnapshot()
  })

  test('add inserted or deleted class to line', () => {
    const result = processHtml(
      dedent`
    <div>
      <pre>
      <code class="language-css">+ p { color: white }
- p { color: red }
! div { background: red }
# div { background: black }
      </code>
      </pre>
    </div>
    `,
      {
        meta: 'diff'
      }
    )
    expect(result).toMatchSnapshot()
  })

  test('option remove work', () => {
    const result = processHtml(
      dedent`   <div>
    <pre>
    <code class="language-css">+ p { color: white }
- p { color: red }
! div { background: red }
# div { background: black }
    </code>
    </pre>
  </div>`,
      { meta: 'diff', optionsDiff: { remove: true } }
    )
    expect(result).toMatchSnapshot()
  })

  test('option classMapping work', () => {
    const result = processHtml(
      dedent`   <div>
    <pre>
    <code class="language-css">+ p { color: white }
- p { color: red }
    </code>
    </pre>
  </div>`,
      { meta: 'diff', optionsDiff: { remove: true, classMapping: { diff: 'custom-diff' } } }
    )
    expect(result).toMatchSnapshot()
  })
})

describe('rehypre-prism-plus plugin', () => {
  const processHtml = (
    html: string,
    setting?: { options?: RehypePrismPlusOptions; optionsDiff?: Options; meta?: string }
  ) => {
    return rehype()
      .data('settings', { fragment: true })
      .use(addMeta, setting?.meta || '')
      .use(rehypePrismPlus, setting?.options)
      .use(rehypePrismDiff, setting?.optionsDiff)
      .processSync(html)
      .toString()
  }

  test('should no diff', () => {
    const result = processHtml(
      dedent`
    <div>
      <pre>
      <code class="language-css">p { color: red }</code>
      </pre>
    </div>
    `
    )
    expect(result).toMatchSnapshot()
  })

  test('should add diff', () => {
    const result = processHtml(
      dedent`
      <div>
        <pre>
        <code class="language-css">p { color: red }</code>
        </pre>
      </div>
    `,
      {
        meta: 'diff'
      }
    )
    expect(result).toMatchSnapshot()
  })

  test('add inserted or deleted class to line', () => {
    const result = processHtml(
      dedent`
    <div>
      <pre>
      <code class="language-css">+ p { color: white }
- p { color: red }
! div { background: red }
# div { background: black }
      </code>
      </pre>
    </div>
    `,
      {
        meta: 'diff'
      }
    )
    expect(result).toMatchSnapshot()
  })

  test('option remove work', () => {
    const result = processHtml(
      dedent`   <div>
    <pre>
    <code class="language-css">+ p { color: white }
- p { color: red }
! div { background: red }
# div { background: black }
    </code>
    </pre>
  </div>`,
      { meta: 'diff', optionsDiff: { remove: true } }
    )
    expect(result).toMatchSnapshot()
  })

  test('option classMapping work', () => {
    const result = processHtml(
      dedent`   <div>
    <pre>
    <code class="language-css">+ p { color: white }
- p { color: red }
    </code>
    </pre>
  </div>`,
      { meta: 'diff', optionsDiff: { remove: true, classMapping: { diff: 'custom-diff' } } }
    )
    expect(result).toMatchSnapshot()
  })
})
