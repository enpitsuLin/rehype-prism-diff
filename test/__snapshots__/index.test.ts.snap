// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rehypre-prism plugin add inserted or deleted class to line 1`] = `
"<div>
  <pre class=\\"language-css\\">  <code class=\\"language-css code-diff\\"><span class=\\"token selector diff-inserted\\">+ p</span> <span class=\\"token punctuation\\">{</span> <span class=\\"token property\\">color</span><span class=\\"token punctuation\\">:</span> white <span class=\\"token punctuation\\">}</span>
<span class=\\"token selector diff-deleted\\">- p</span> <span class=\\"token punctuation\\">{</span> <span class=\\"token property\\">color</span><span class=\\"token punctuation\\">:</span> red <span class=\\"token punctuation\\">}</span>
  </code>
  </pre>
</div>"
`;

exports[`rehypre-prism plugin should add diff 1`] = `
"<div>
  <pre class=\\"language-css\\">  <code class=\\"language-css code-diff\\"><span class=\\"token selector\\">p</span> <span class=\\"token punctuation\\">{</span> <span class=\\"token property\\">color</span><span class=\\"token punctuation\\">:</span> red <span class=\\"token punctuation\\">}</span></code>
  </pre>
</div>"
`;

exports[`rehypre-prism-plus plugin add inserted or deleted class to line 1`] = `
"<div>
  <pre class=\\"language-css\\">  <code class=\\"language-css code-highlight code-diff\\"><span class=\\"code-line diff-inserted\\"><span class=\\"token selector\\"><span class=\\"token combinator\\">+</span> p</span> <span class=\\"token punctuation\\">{</span> <span class=\\"token property\\">color</span><span class=\\"token punctuation\\">:</span> <span class=\\"token color\\">white</span> <span class=\\"token punctuation\\">}</span>
</span><span class=\\"code-line diff-deleted\\"><span class=\\"token selector\\">- p</span> <span class=\\"token punctuation\\">{</span> <span class=\\"token property\\">color</span><span class=\\"token punctuation\\">:</span> <span class=\\"token color\\">red</span> <span class=\\"token punctuation\\">}</span>
</span></code>
  </pre>
</div>"
`;

exports[`rehypre-prism-plus plugin should add diff 1`] = `
"<pre class=\\"language-css\\"><code class=\\"language-css code-highlight code-diff\\"><span class=\\"code-line\\"><span class=\\"token selector\\">p</span> <span class=\\"token punctuation\\">{</span> <span class=\\"token property\\">color</span><span class=\\"token punctuation\\">:</span> <span class=\\"token color\\">red</span> <span class=\\"token punctuation\\">}</span></span></code>
</pre>"
`;
